### PR TITLE
Fix TorchLayer output shape for batched tuple expval returns

### PR DESCRIPTION
**Pull request by dcasota**
TorchLayer._evaluate_qnode produced shape (n_measurements, batch, 1) instead of (batch, n_measurements) when the QNode returns multiple measurements as a tuple and receives batched inputs. The PR #5816 hotfix wrapped each tensor result individually via tuple(_combine_dimensions([r]) for r in res), returning a tuple that forward() then stacked along dim=0 producing the wrong layout.

Fix by passing all tensor results as a single list to _combine_dimensions, which reshapes each to (batch, 1) and hstacks them into a single (batch, n_measurements) tensor. This avoids the tuple path in forward() entirely, producing the correct output shape for downstream nn.Linear layers.

Also handle list returns (not just tuple) for robustness, and remove the pragma: no cover annotation since the code path is now properly tested.

Assisted-by: Droid:Claude:claude-opus-4-6

### Checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

 **Context:**

`qml.qnn.TorchLayer` produces incorrect output shape when the wrapped QNode returns multiple measurements as a tuple
(e.g., `return qml.expval(qml.Z(0)), qml.expval(qml.Z(1))`) and receives batched inputs. This is the
natural/documented pattern for multi-output quantum circuits, and it breaks when used inside `torch.nn.Sequential`
with downstream `nn.Linear` layers.

PR #5816 previously added a hotfix for a related issue, but the fix in `_evaluate_qnode` wraps each tensor result
individually via `tuple(_combine_dimensions([r]) for r in res)`, producing a tuple of `(batch, 1)` tensors. The
`forward()` method then calls `torch.stack(results, dim=0)`, yielding shape `(n_measurements, batch, 1)` instead of
the expected `(batch, n_measurements)`.

**Description of the Change:**

Two-line change in `TorchLayer._evaluate_qnode`:

1. Broadened the `isinstance` check from `tuple` to `(tuple, list)` for robustness.
2. When all elements in the result are tensors, pass them as a single flat list to `_combine_dimensions` instead of
wrapping each individually. This reshapes each `(batch,)` tensor to `(batch, 1)` and hstacks them into a single
`(batch, n_measurements)` tensor, which avoids the tuple code path in `forward()` entirely.

Removed the `# pragma: no cover` annotation since the code path is now properly tested.

Added a regression test `test_forward_tuple_batched_expval` that verifies:
- Output shape is `(5, 2)` not `(2, 5, 1)` for 5-sample batch with 2 expval measurements
- Gradient flows correctly through the layer to all parameters

**Benefits:**

- QNodes returning tuples of `qml.expval` now produce correct `(batch, n_measurements)` output shape with batched
inputs
- TorchLayer can be used in `nn.Sequential` pipelines with downstream `nn.Linear` layers without shape mismatch errors
- The previously untested code path (`# pragma: no cover`) is now covered by a regression test

**Possible Drawbacks:**

- Code that relied on the previous `(n_measurements, batch, 1)` output shape (a transposed layout) would need to be
updated. However, this shape was never documented or intended, and it was incompatible with standard PyTorch
`nn.Linear` usage, so such code was likely working around the bug already.

**Related GitHub Issues:**

- Related to PR #5816 which added the initial hotfix for TorchLayer tuple returns

Assisted-by: Droid:Claude:claude-opus-4-6

### CODE REVIEW

**Comment by dcasota** [comment]
@JerryChen97 Could the team also review this PR for a future release?

**Comment by Alex-Preciado** [comment]
Hi @dcasota , thanks for opening this PR 🚀! We are currently heads-down preparing the upcoming PennyLane release (next week), but we will make sure to find a reviewer for a first pass this week. In the meantime, please let us know if you have any questions!

File: tests/qnn/test_qnn_torch.py [hunk 1]
```diff
@@ -600,6 +600,44 @@ def qnode(inputs, weights):
...
+    assert result.shape == (5, 2), (
+        f"Expected shape (5, 2) but got {result.shape}. "
+        "TorchLayer must produce (batch_size, n_measurements) for tuple expval returns."
+    )
+
+    # Verify gradient flow through the layer
+    loss = result.sum()
+    loss.backward()
+    for param in qlayer.parameters():
+        assert param.grad is not None, "Gradient did not flow to TorchLayer weights"
```
Comment (astralcai) [inline tests/qnn/test_qnn_torch.py]: Please replace all references to `qml` with `qp` here.

File: doc/releases/changelog-0.45.0.md [hunk 2]
```diff
@@ -1228,6 +1228,13 @@
 
 <h3>Bug fixes 🐛</h3>
 
+* Fixed a bug where `qp.qnn.TorchLayer` produced incorrect output shape `(n_measurements, batch, 1)`
+  instead of `(batch, n_measurements)` when the wrapped QNode returns multiple measurements as a tuple
+  (e.g., `return qp.expval(qp.Z(0)), qp.expval(qp.Z(1))`) and receives batched inputs. This
+  previously caused shape mismatch errors when feeding the output into downstream `torch.nn.Linear`
+  layers.
+  [(#9284)](https://github.com/PennyLaneAI/pennylane/pull/9284)
```
Comment (astralcai) [inline doc/releases/changelog-0.45.0.md]: Can you move this to `doc/releases/changelog-dev.md`?

**Comment by dcasota** [comment]
Hi @astralcai, 

I've processed both changes as suggested.

With regard to the failed documentation tests: The CI step `pytest doc/releases/changelog-dev.md` fails with pytest exit code 5 ("no tests collected") because the empty `changelog-dev.md` template introduced by #9408 contains no Sybil-collectable content - no >>> doctest sessions, fenced ```python blocks, or .. code-block:: python directives - and any PR touching the dev changelog inherits this until the file is seeded.

Workflow-level workarounds (tolerating exit 5, continue-on-error, or dropping the dedicated step) require write access on PennyLaneAI/pennylane` that external contributors don't have. 

In my opinion, the cleanest fix is a tiny maintainer-side seeding PR adding one trivial doctest or fenced code block to `changelog-dev.md` on main, which clears the failure for every in-flight PR at once without conflating scope into unrelated bug-fix branches.

Hope this helps.
Daniel

**Comment by JerryChen97** [comment]
> Hi @astralcai,
> 
> I've processed both changes as suggested.
> 
> With regard to the failed documentation tests: The CI step `pytest doc/releases/changelog-dev.md` fails with pytest exit code 5 ("no tests collected") because the empty `changelog-dev.md` template introduced by #9408 contains no Sybil-collectable content - no >>> doctest sessions, fenced ```python blocks, or .. code-block:: python directives - and any PR touching the dev changelog inherits this until the file is seeded.
> 
> Workflow-level workarounds (tolerating exit 5, continue-on-error, or dropping the dedicated step) require write access on PennyLaneAI/pennylane` that external contributors don't have.
> 
> In my opinion, the cleanest fix is a tiny maintainer-side seeding PR adding one trivial doctest or fenced code block to `changelog-dev.md` on main, which clears the failure for every in-flight PR at once without conflating scope into unrelated bug-fix branches.
> 
> Hope this helps. Daniel

Hi @dcasota per the doctest failure please feel free to ignore it~ It's a rather new system and we are constantly improving still. And this CI so far has not yet been a necessary requirement; it's mostly serving as reference.
I've already noticed this issue and had it noted internally! Hopefully very soon we would also have enough bandwidth to resolve this entirely. Thanks for the thorough analysis and suggestion anyways!

**Comment by Alex-Preciado** [comment]
> With regard to the failed documentation tests: The CI step `pytest doc/releases/changelog-dev.md` fails with pytest exit code 5 ("no tests collected") because the empty `changelog-dev.md` template introduced by #9408 contains no Sybil-collectable content - no >>> doctest sessions, fenced ```python blocks, or .. code-block:: python directives - and any PR touching the dev changelog inherits this until the file is seeded.
> 
> Workflow-level workarounds (tolerating exit 5, continue-on-error, or dropping the dedicated step) require write access on PennyLaneAI/pennylane` that external contributors don't have.
> 
> In my opinion, the cleanest fix is a tiny maintainer-side seeding PR adding one trivial doctest or fenced code block to `changelog-dev.md` on main, which clears the failure for every in-flight PR at once without conflating scope into unrelated bug-fix branches.
> 
> Hope this helps. Daniel

Hi @dcasota, thanks for raising this here. We are actually working on this right now. @andrijapau is preparing a fix in #9411, We're aiming to get this merged very soon. Feel free to follow the progress there 🚀

**Comment by dcasota** [comment]
@Alex-Preciado Thank you for the quick progress on this. Just to confirm - PR #9411 fixes the pytest exit code 5 failure in the documentation workflow that is currently blocking the merge here. Once #9411 is merged, this PR should be fine as well.

**Comment by astralcai** [comment]
Hi @dcasota,

Thank you for your contribution! The PR looks good overall. Please don't forget to add your name to the **Contributors** section of the changelog as well!

Note that a feature freeze is currently in effect as we prepare for the upcoming release, so we won't actually be able to merge anything into the main branch until the new release comes out next week. We will keep you updated.

**Comment by dcasota** [comment]
@astralcai v0.45.0 is not yet released, and the team is heavily working on last features/changes like PR #9277. Luckily the cut has happened, changelog-dev.md is re-headed and main is now accepting 0.46-targeted PRs. Could you have another look this PR?

**Comment by astralcai** [comment]
> @astralcai v0.45.0 is not yet released, and the team is heavily working on last features/changes like PR #9277. Luckily the cut has happened, changelog-dev.md is re-headed and main is now accepting 0.46-targeted PRs. Could you have another look this PR?

Hi @dcasota, the PR that you mentioned made changes that are exclusive to `pennylane/labs`, which is a prototyping ground that is intentionally kept isolated from core source code. Nothing outside of `labs` depend on anything in `labs`, which is why things get merged before the release. For PRs that target actual source code, even though there is a split between the RC branch and the main branch, we still prefer to enforce the merge freeze.

The top priority of the team is currently to prepare for the release, so there is not enough review capacity for external PRs. Please be patient, we will make sure to have someone take a look at this as soon as the release process is concluded.

**Review by astralcai — APPROVED** [review]
Looks good! We'll find you a second reviewer very soon. 🚀

**Review by albi3ro — APPROVED** [review]
Thanks for the fix! looks great :)
